src/indicators/money_flow_index.rs

```diff
@@ -0,0 +1,235 @@
+///
+/// _tp_ - typical price
+/// _rmf_ - raw MF
+/// _pmf_ - positive MF
+/// _nmf_ - negative MF
+///
+/// # Parameters
+///
+/// * _n__ - number of periods, integer greater than 0
+///
+/// # Example
+///
+/// ```
+/// use ta::indicators::MoneyFlowIndex;
+/// use ta::{Next, DataItem};
+///
+/// let mut mfi = MoneyFlowIndex::new(3).unwrap();
+/// let di = DataItem::builder()
+///             .high(3.0)
+///             .low(1.0)
+///             .close(2.0)
+///             .open(1.5)
+///             .volume(1000.0)
+///             .build().unwrap();
+/// mfi.next(&di);
+///
+/// ```
+/// # Links
+/// ![Moeny Flow Index, Wikipedia](https://en.wikipedia.org/wiki/Money_flow_index)
+/// ![Money Flow Index, stockcharts] (https://stockcharts.com/school/doku.php?id=chart_school:technical_indicators:money_flow_index_mfi)
+
+#[derive(Debug, Clone)]
+pub struct MoneyFlowIndex {
+    n: u32,
+    mf_vec: VecDeque<f64>,
+    prev_tp: f64,
+}
+
+impl MoneyFlowIndex {
+    pub fn new(n: u32) -> Result<Self> {
+        match n {
+            0 => Err(Error::from_kind(ErrorKind::InvalidParameter)),
+            _ => {
+                let indicator = Self {
+                    n: n,
+                    mf_vec: VecDeque::with_capacity(n as usize + 1),
+                    prev_tp: 0.0,
+                };
+                Ok(indicator)
+            }
+        }
+    }
+}
+
+impl Next<f64> for MoneyFlowIndex {
+    type Output = f64;
+
+    fn next(&mut self, mf: f64) -> f64 {
+        if (self.mf_vec.len() as u32) < self.n {
+            self.mf_vec.push_back(mf);
+        } else {
+            self.mf_vec.pop_front();
+            self.mf_vec.push_back(mf);
+        }
+
+        if self.mf_vec.len() as u32 == self.n {
+            let pmf: f64 = self.mf_vec.iter().filter(|&i| *i >= 0.0).sum();
+            let nmf: f64 = self.mf_vec.iter().filter(|&i| *i < 0.0).sum();
+            ((pmf / (pmf + nmf.abs())) * 100.0)
+        } else {
+            0.0
+        }
+    }
+}
+
+impl<'a, T: High + Low + Close + Volume> Next<&'a T> for MoneyFlowIndex {
+    type Output = f64;
+
+    fn next(&mut self, input: &'a T) -> f64 {
+        // self.next((input.high(), input.low(), input.close(), input.volume()))
+
+        let high = input.high();
+        let low = input.low();
+        let close = input.close();
+        let volume = input.volume();
+
+        let tp = (high + low + close) / 3.0;
+        let rmf = tp * volume;
+        let mf = if tp >= self.prev_tp {
+            rmf
+        } else {
+            (rmf * -1.0)
+        };
+
+        self.prev_tp = tp;
+
+        if (self.mf_vec.len() as u32) < self.n {
+            self.mf_vec.push_back(mf);
+        } else {
+            self.mf_vec.pop_front();
+            self.mf_vec.push_back(mf);
+        }
+
+        if self.mf_vec.len() as u32 == self.n {
+            let pmf: f64 = self.mf_vec.iter().filter(|&i| *i >= 0.0).sum();
+            let nmf: f64 = self.mf_vec.iter().filter(|&i| *i < 0.0).sum();
```

Comment:
Iterating through all saved items every time a new item comes, makes the complexity of this indicator `O(N)`.
Most of the indicators have complexity `O(1)` and I believe it can be achieved here as well. 

How: instead of calculating sum every time we can just keep it as a field and adjust when we do `pop_front` and `push_back` (if incoming `mf` is negative, we should it convert into positive, so `sum` always statys positive`). We need also to track `pmf`  - that is enough (`nmf` is not needed actually).
With every `pop_front` we would also adjust `pmf` if the value that we're taking out is positive.

Does it make sense?

Reply:
Changed implementation as stated above in new commit.
Please let me know in case more changes are required